conctr.agent.lib.nut

```diff
@@ -195,18 +232,296 @@ class Conctr {
     }
 
 
-    //
+    // 
+    // Sets the protocal that should be used 
+    // 
+    // @param   {String}    protocol Either amqp or mqtt
+    // @return  {String}    current protocal after change
+    // 
+    function setProtocol(protocol) {
+        if (protocol == AMQP || protocol == MQTT) {
+            _protocol = protocol;
+        } else {
+            server.error(protocol + " is not a valid protocol.");
+        }
+        _pubSubEndpoints = _formPubSubEndpointUrls(_app_id, _api_key, _device_id, _region, _env);
+        return _protocol
+    }
+
+
+    // 
     // Requests the device sends its current location to the agent or to Conctr
-    //
+    // 
     // @param  {Boolean} sendToConctr - If true the location will be sent to Conctr. If false, it will be cached on the agent.
-    //
+    // 
     function sendLocation(sendToConctr = true) {
 
         if (DEBUG) server.log("Conctr: requesting location be sent to " + (sendToConctr ? "conctr" : "agent"));
         _sender.send(LOCATION_REQ_EVENT, sendToConctr);
 
     }
-    
+
+
+    // 
+    // Publishes a message to a specific topic.
+    // 
+    // @param  {String/Array}   topics      List of Topics that message should be sent to
+    // @param  {[type]}         msg         Data to be sent to be published
+    // @param  {[String}        contentType Header specifying the content type of the msg
+    // @param  {Function}       cb          Function called on completion of publish request
+    // 
+    function publish(topics, msg, contentType = null, cb = null) {
+        local relativeUrl = ""
+        _publish(relativeUrl, { "topics": topics }, msg, contentType, cb);
+
+    }
+
+
+    // 
+    // Publishes a message to a specific device.
+    // @param  {String/Array}   deviceId     Device id(s) the message should be sent to
+    // @param  {[type]}         msg          Data to be sent to be published
+    // @param  {[type]}         contentType  Header specifying the content type of the msg
+    // @param  {Function}       cb           Function called on completion of publish request
+    // 
+    function publishToDevice(deviceIds, msg, contentType = null, cb = null) {
```

Comment:
What happens if they provide a callback but no contentType?

Reply:
child function called handles the switching